statsmodels/genmod/tests/test_glm_weights.py

```diff
@@ -182,3 +221,367 @@ def setupClass(cls):
+
+        # Need to make a few adjustments...
+        # First, calculate scale using nobs as denominator
+        scale = res1.scale * model.df_resid / model.wnobs
+        # Calculate llf using adj scale and wts = freq_weights
+        wts = model.freq_weights
+        llf = model.family.loglike(model.endog, res1.mu, wts, scale)
+        # SM uses (essentially) stat's loglike formula... first term is
+        # (endog - mu) ** 2 / scale
+        adj_sm = -1 / 2 * ((model.endog - res1.mu) ** 2).sum() / scale
+        # R has these 2 terms that stata/sm don't
+        adj_r = -model.wnobs / 2 + np.sum(np.log(model.var_weights)) / 2
+        llf_adj = llf - adj_sm + adj_r
+        assert_allclose(llf_adj, res2.ll, atol=1e-6, rtol=1e-7)
+
+
+def gen_endog(lin_pred, family_class, link, binom_version=0):
+
+    np.random.seed(872)
+
+    fam = sm.families
+
+    mu = link().inverse(lin_pred)
+
+    if family_class == fam.Binomial:
+        if binom_version == 0:
+            endog = 1*(np.random.uniform(size=len(lin_pred)) < mu)
+        else:
+            endog = np.empty((len(lin_pred), 2))
+            n = 10
+            endog[:, 0] = (np.random.uniform(size=(len(lin_pred), n)) < mu[:, None]).sum(1)
+            endog[:, 1] = n - endog[:, 0]
+    elif family_class == fam.Poisson:
+        endog = np.random.poisson(mu)
+    elif family_class == fam.Gamma:
+        endog = np.random.gamma(2, mu)
+    elif family_class == fam.Gaussian:
+        endog = mu + np.random.normal(size=len(lin_pred))
+    elif family_class == fam.NegativeBinomial:
+        from scipy.stats.distributions import nbinom
+        endog = nbinom.rvs(mu, 0.5)
+    elif family_class == fam.InverseGaussian:
+        from scipy.stats.distributions import invgauss
+        endog = invgauss.rvs(mu)
+    else:
+        raise ValueError
+
+    return endog
+
+
+def test_wtd_gradient_irls():
+    # Compare the results when using gradient optimization and IRLS.
+    # TODO: Find working examples for inverse_squared link
+
+    np.random.seed(87342)
+
+    fam = sm.families
+    lnk = sm.families.links
+    families = [(fam.Binomial, [lnk.logit, lnk.probit, lnk.cloglog, lnk.log,
+                                lnk.cauchy]),
+                (fam.Poisson, [lnk.log, lnk.identity, lnk.sqrt]),
+                (fam.Gamma, [lnk.log, lnk.identity, lnk.inverse_power]),
+                (fam.Gaussian, [lnk.identity, lnk.log, lnk.inverse_power]),
+                (fam.InverseGaussian, [lnk.log, lnk.identity,
+                                       lnk.inverse_power,
+                                       lnk.inverse_squared]),
+                (fam.NegativeBinomial, [lnk.log, lnk.inverse_power,
+                                        lnk.inverse_squared, lnk.identity])]
+
+    n = 100
+    p = 3
+    exog = np.random.normal(size=(n, p))
+    exog[:, 0] = 1
+
+    skip_one = False
+    for family_class, family_links in families:
+        for link in family_links:
```

Comment:
Is this part mostly copied from another test?

(just a question for now, but we might want to eventually reuse code.)